### PR TITLE
Waveform: fix shifted (text) markers whe using odd scale factors

### PR DESCRIPTION
**Pull request by ronso0**
https://bugs.launchpad.net/mixxx/+bug/1917253

set an non-int scale factor like
`export QT_SCREEN_SCALE_FACTORS=1.37`
in the images below all three markers are actually exactly one the play position.

before
image

this PR
image

### CODE REVIEW

**Comment by ronso0** [comment]
tbh I didn't follow all those confusing scale functions. so consider this a quick fix for the skins that are using text markers.

there's still something wrong with markers that use icons (Deere only, at least in 2.3):
range is correct, markers are shifted
image

**Comment by ronso0** [comment]
...and that appearantly depends on the icon width: the wider, the more it is shifted.
dunno how to fix it...the image shift function works, maybe there's an offset problem when pixmaps are loaded
https://github.com/mixxxdj/mixxx/blob/2.3/src/waveform/renderers/waveformrendermark.cpp#L149-L161

**Comment by daschuer** [comment]
I can confirm the issue and the fix with 3 / 2.5 / 2 / 1.5 / 1: 
before 
grafik
after
grafik

On Ubuntu Groovy

File: src/waveform/renderers/waveformrendermark.cpp [hunk 1]
```diff
@@ -146,6 +146,7 @@ void WaveformRenderMark::slotCuesUpdated() {
 
 void WaveformRenderMark::generateMarkImage(WaveformMarkPointer pMark) {
     // Load the pixmap from file -- takes precedence over text.
+    float devicePixelRatio = m_waveformRenderer->getDevicePixelRatio();
     if (!pMark->m_pixmapPath.isEmpty()) {
         QString path = pMark->m_pixmapPath;
         QImage image = *WImageStore::getImage(path, scaleFactor());
```
Comment (daschuer) [inline src/waveform/renderers/waveformrendermark.cpp:152]: Here scaleFactor() is in use. 
Which value has it?  
I did not understand why scaleFactor() works here. 
Does it? 
When is the code used?

Comment (ronso0) [reply src/waveform/renderers/waveformrendermark.cpp:152]: scaleFactor or devicePixelRatio makes no difference here.
It's 1.0 regardless of the Qt scale factor. some leftover from custom Mixxx scaling? 
this is used when markers use icons, like in Deere (Cue, loop_in/out).
then the entire part below is skipped
= the icon needs to be centered so the icon stroke (if any) aligns with the marker position on the waveform

also, the one initial problem here causing the offset is that the marker sroke is not taken from the samplePosition but directly.

Comment (ronso0) [reply src/waveform/renderers/waveformrendermark.cpp:152]: > I did not understand why scaleFactor() works here.

must admit I don't understand it either.
The fixes apply to all QPainter-related calculations below.

Comment (daschuer) [reply src/waveform/renderers/waveformrendermark.cpp:152]: Deere uses it for its cue marker. 
I can confirm that it scales fine.

**Review by daschuer — APPROVED** [review]
LGTM, Thnak you.

**Comment by ronso0** [comment]
> ...and that appearantly depends on the icon width: the wider, the more it is shifted.
> dunno how to fix it...the image shift function works, maybe there's an offset problem when pixmaps are loaded
> https://github.com/mixxxdj/mixxx/blob/2.3/src/waveform/renderers/waveformrendermark.cpp#L149-L161

got it!
#3938

**Pull request merged**